### PR TITLE
Line ending test

### PR DESCRIPTION
**Pull request by bpbond**
Ensure that all CSV input files (whether compressed or not) have correct line ending encoding and end with a blank final line. Addresses #867

### CODE REVIEW

**Comment by pralitp** [comment]
The first check: re-running the `normalize_files("./inst/")` on a clean checkout does not produce any differences. I.e. we don't need to put special checks to avoid re-zipping if the underlying data didn't change.

**Comment by bpbond** [comment]
Sweet! God, I write such good code 😆

**Comment by pralitp** [comment]
Second check: Modify CSV files with excel with difference save options
```
$ # Verify LF line endings
WE34383:energy pralitp$ vi A22.sector.csv 
$ # Open in Excel, no changes, Save using default settings
WE34383:energy pralitp$ open A22.sector.csv 
WE34383:energy pralitp$ git status
On branch master
Your branch is up-to-date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   A22.sector.csv

no changes added to commit (use "git add" and/or "git commit -a")
$ # verify Windows (CRLF) line endings
$ vi A22.sector.csv
$ # git diff seems to mostly handle it well, not sure there is anything we can do about the
$ # added trailing commas
$ # Note the remarks about it wanting to covert CRLF to LF and no line ending at the end
WE34383:energy pralitp$ git diff
warning: CRLF will be replaced by LF in inst/extdata/energy/A22.sector.csv.
The file will have its original line endings in your working directory.
diff --git a/inst/extdata/energy/A22.sector.csv b/inst/extdata/energy/A22.sector.csv
index aac8c7bc..2e50d3a3 100644
--- a/inst/extdata/energy/A22.sector.csv
+++ b/inst/extdata/energy/A22.sector.csv
@@ -1,9 +1,9 @@
-# File: A22.sector.csv
-# Title: Primary energy transformation supplysectors
-# Units: N/A
-# Description: Energy transformation supplysector units and logit exponents
+# File: A22.sector.csv,,,,,
+# Title: Primary energy transformation supplysectors,,,,,
+# Units: N/A,,,,,
+# Description: Energy transformation supplysector units and logit exponents,,,,,
 supplysector,output.unit,input.unit,price.unit,logit.exponent,logit.type
 gas processing,EJ,EJ,1975$/GJ,-6,
 nuclearFuelGenII,EJ,EJ,1975$/GJ,-3,
 nuclearFuelGenIII,EJ,EJ,1975$/GJ,-3,
-refining,EJ,EJ,1975$/GJ,-6,absolute-cost-logit
+refining,EJ,EJ,1975$/GJ,-6,absolute-cost-logit
\ No newline at end of file
$ # run normalize_files("./inst/") in R
$ # verify LF endings
$ vi A22.sector.csv
$ # now we get a clean diff (except for trailing commas)
WE34383:energy pralitp$ git diff .
diff --git a/inst/extdata/energy/A22.sector.csv b/inst/extdata/energy/A22.sector.csv
index aac8c7bc..e6106a3d 100644
--- a/inst/extdata/energy/A22.sector.csv
+++ b/inst/extdata/energy/A22.sector.csv
@@ -1,7 +1,7 @@
-# File: A22.sector.csv
-# Title: Primary energy transformation supplysectors
-# Units: N/A
-# Description: Energy transformation supplysector units and logit exponents
+# File: A22.sector.csv,,,,,
+# Title: Primary energy transformation supplysectors,,,,,
+# Units: N/A,,,,,
+# Description: Energy transformation supplysector units and logit exponents,,,,,
 supplysector,output.unit,input.unit,price.unit,logit.exponent,logit.type
 gas processing,EJ,EJ,1975$/GJ,-6,
 nuclearFuelGenII,EJ,EJ,1975$/GJ,-3,
$ # open again in Excel, this time Save As Mac OS csv (default in older versions of Excel on Mac)
WE34383:energy pralitp$ open A22.sector.csv 
$ # verify CR line endings
WE34383:energy pralitp$ vi A22.sector.csv 
$ # git diff chokes on this and no messages indicating it wants to help us convert line endings
WE34383:energy pralitp$ git diff
diff --git a/inst/extdata/energy/A22.sector.csv b/inst/extdata/energy/A22.sector.csv
index aac8c7bc..d915fdf6 100644
--- a/inst/extdata/energy/A22.sector.csv
+++ b/inst/extdata/energy/A22.sector.csv
@@ -1,9 +1 @@
-# File: A22.sector.csv
-# Title: Primary energy transformation supplysectors
-# Units: N/A
-# Description: Energy transformation supplysector units and logit exponents
-supplysector,output.unit,input.unit,price.unit,logit.exponent,logit.type
-gas processing,EJ,EJ,1975$/GJ,-6,
-nuclearFuelGenII,EJ,EJ,1975$/GJ,-3,
-nuclearFuelGenIII,EJ,EJ,1975$/GJ,-3,
-refining,EJ,EJ,1975$/GJ,-6,absolute-cost-logit
+# File: A22.sector.csv,,,,,^M# Title: Primary energy transformation supplysectors,,,,,^M# Units: N/A,,,,,^M# Description: Energy transformation supplysector units and logit exponents,,,,,^Msupplysector,output.unit,input.unit,price.unit,logit.exponent,logit.type^Mgas processing,EJ,EJ,1975$/GJ,-6,^MnuclearFuelGenII,EJ,EJ,1975$/GJ,-3,^MnuclearFuelGenIII,EJ,EJ,1975$/GJ,-3,^Mrefining,EJ,EJ,1975$/GJ,-6,absolute-cost-logit
\ No newline at end of file
$ # run normalize_files("./inst/") in R
$ # now we get a clean diff (except for trailing commas)
WE34383:energy pralitp$ git diff
diff --git a/inst/extdata/energy/A22.sector.csv b/inst/extdata/energy/A22.sector.csv
index aac8c7bc..e6106a3d 100644
--- a/inst/extdata/energy/A22.sector.csv
+++ b/inst/extdata/energy/A22.sector.csv
@@ -1,7 +1,7 @@
-# File: A22.sector.csv
-# Title: Primary energy transformation supplysectors
-# Units: N/A
-# Description: Energy transformation supplysector units and logit exponents
+# File: A22.sector.csv,,,,,
+# Title: Primary energy transformation supplysectors,,,,,
+# Units: N/A,,,,,
+# Description: Energy transformation supplysector units and logit exponents,,,,,
 supplysector,output.unit,input.unit,price.unit,logit.exponent,logit.type
 gas processing,EJ,EJ,1975$/GJ,-6,
 nuclearFuelGenII,EJ,EJ,1975$/GJ,-3,
```

**Comment by pralitp** [comment]
First check again, this time on Windows (and I've have removed the not on Windows check in `normalize_files`):
running the normalize_files("./inst/") on a clean checkout does not produce any differences in the just .csv files however the all the .gz files do differ.
I think this is because: 

1. When R writes the file back out it writes it using the native line endings
2. Git is converting the .csv files to CRLF automatically on checkout (and generally is willing to make conversions between CRLF and LF) because we have `.gitattribtues`: `* text=auto`
3. Git isn't going to do anything about line endings in a .gz file.  Thus when R re-writes the files on windows the CRLF gets locked into the .gz file.

**Comment by pralitp** [comment]
Second check, this time on Windows:
```
$ # verify CRLF line endings
$ vi A22.sector.csv
$ # open in excel and Save unmodified
$ explorer A22.sector.csv
$ # verify line endings are still CRLF
$ vi A22.sector.csv
$ # diff shows up clean (except for trailing commas)
$ # not even any messages about line ending conversions or missing one at the end
$ git diff A22.sector.csv
diff --git a/inst/extdata/energy/A22.sector.csv b/inst/extdata/energy/A22.sector.csv
index aac8c7b..e6106a3 100644
--- a/inst/extdata/energy/A22.sector.csv
+++ b/inst/extdata/energy/A22.sector.csv
@@ -1,7 +1,7 @@
-# File: A22.sector.csv
-# Title: Primary energy transformation supplysectors
-# Units: N/A
-# Description: Energy transformation supplysector units and logit exponents
+# File: A22.sector.csv,,,,,
+# Title: Primary energy transformation supplysectors,,,,,
+# Units: N/A,,,,,
+# Description: Energy transformation supplysector units and logit exponents,,,,,
 supplysector,output.unit,input.unit,price.unit,logit.exponent,logit.type
 gas processing,EJ,EJ,1975$/GJ,-6,
 nuclearFuelGenII,EJ,EJ,1975$/GJ,-3,
$ # run normalize_files("./inst/") in R
$ # verify CRLF line endings
$ vi A22.sector.csv
$ # still clean diff (except for trailing commas)
$ git diff A22.sector.csv
diff --git a/inst/extdata/energy/A22.sector.csv b/inst/extdata/energy/A22.sector.csv
index aac8c7b..e6106a3 100644
--- a/inst/extdata/energy/A22.sector.csv
+++ b/inst/extdata/energy/A22.sector.csv
@@ -1,7 +1,7 @@
-# File: A22.sector.csv
-# Title: Primary energy transformation supplysectors
-# Units: N/A
-# Description: Energy transformation supplysector units and logit exponents
+# File: A22.sector.csv,,,,,
+# Title: Primary energy transformation supplysectors,,,,,
+# Units: N/A,,,,,
+# Description: Energy transformation supplysector units and logit exponents,,,,,
 supplysector,output.unit,input.unit,price.unit,logit.exponent,logit.type
 gas processing,EJ,EJ,1975$/GJ,-6,
 nuclearFuelGenII,EJ,EJ,1975$/GJ,-3,
```

**Comment by pralitp** [comment]
@bpbond, would it be easy to force it to write LF anytime it is going to write .gz?

**Comment by pralitp** [comment]
Ok, I've got some updates here but not quite a solution.  Your latest push had a bug but if you always force LF all the files show up modified on Windows but a git diff will indicate no differences (with warnings about wanting to convert line endings).

I've figured out an easy work around: force LF only when we are going to compress.  But even then the `.gz` files show up modified.  I can verify the file inside of the `.gz` are LF and identical once extracted.

Converting the `.gz` files to hex and doing a diff indicates there is only a difference at the very start of the file so maybe it is some setting we need to change:
```
1c1
< 00000000: 1f8b 0800 0000 0000 0003 ecfd 5b93 24c7  ............[.$.
---
> 00000000: 1f8b 0800 0000 0000 0006 ecfd 5b93 24c7  ............[.$.
```

Anyways there is where I am going to have to leave it for the weekend.

**Comment by pralitp** [comment]
A follow up to my last post, `.gz` files have an OS flag baked into the header of the archive:
http://www.onicos.com/staff/iz/formats/gzip.html

I couldn't find any way to make it stop so I went with the best I could come up with: overwrite the flag after the fact to always be `Unix`.

File: DESCRIPTION [hunk 1]
```diff
@@ -24,12 +24,10 @@ Authors@R: c(person("Ben", "Bond-Lamberty",
   person("Rachel", "Hoesly",      role = "ctb"),
   person("Mohamad", "Hejazi",     role = "ctb"),
   person("Gokul", "Iyer",         role = "ctb"),
-  person("Sonny", "Kim",          role = "ctb"),
```
Comment (pralitp) [inline DESCRIPTION]: Double checking if this was intended or not

**Comment by bpbond** [comment]
On the one hand, it seems to work, kudos. On the other...jeez, mucking around in the gzip header is way down in the weeds, and makes me question whether it's worth it.
* Is it worth it to support compressed files? We could remove a lot of code if we said everything in `inst/extdata` had to be uncompressed.
* Is it worth it to worry about touching files when running `normalize_files`? If it's going to be run frequently, yes, but if it's more of a `git bfg` tool, run very rarely, then not. Is it better to simply provide a `normalize_file` routine that fixes a single file (and not support directory-wide normalizing)?

**Comment by bpbond** [comment]
Multiple failures happening in `test_file_encoding` now.

**Comment by bpbond** [comment]
```
───────────────────────────────────────────────────────────────────────────────
test_file_encoding.R:21: failure: inputs are encoded correctly
as.vector(tools::md5sum(f)) not equal to as.vector(tools::md5sum(new_f)).
1/1 mismatches
x[1]: "635076f420b2e24c682f6da85259d7ef"
y[1]: "b611b637bd7dca696b661c719fbb0dce"
Normalized file does not match for  ../../inst/extdata/aglu/LDS/Land_type_area_ha.csv.gz

test_file_encoding.R:21: failure: inputs are encoded correctly
as.vector(tools::md5sum(f)) not equal to as.vector(tools::md5sum(new_f)).
1/1 mismatches
x[1]: "849d567796cb83bc04d72aac6cf03767"
y[1]: "12369c0a05c7435c709eb8fec1f7df13"
Normalized file does not match for  ../../inst/extdata/aglu/LDS/LDS_ag_HA_ha.csv.gz

test_file_encoding.R:21: failure: inputs are encoded correctly
as.vector(tools::md5sum(f)) not equal to as.vector(tools::md5sum(new_f)).
1/1 mismatches
x[1]: "1cb1a733bfba06835aec6560145fdc27"
y[1]: "545e6a5e7f5c1e215c7ad3a25d93e384"
Normalized file does not match for  ../../inst/extdata/aglu/LDS/LDS_ag_prod_t.csv.gz

test_file_encoding.R:21: failure: inputs are encoded correctly
as.vector(tools::md5sum(f)) not equal to as.vector(tools::md5sum(new_f)).
1/1 mismatches
x[1]: "eab51a2ef7a950d2a1e77ff2ead0b466"
y[1]: "3439256bb497ddcb9c67e443af217816"
Normalized file does not match for  ../../inst/extdata/aglu/LDS/Mueller_yield_levels.csv.gz

test_file_encoding.R:21: failure: inputs are encoded correctly
as.vector(tools::md5sum(f)) not equal to as.vector(tools::md5sum(new_f)).
1/1 mismatches
x[1]: "6473f0727d10a640fa9b7285526d9d2b"
y[1]: "855203ab456ac7943286a1c5d4bb2142"
Normalized file does not match for  ../../inst/extdata/aglu/LDS/Water_footprint_m3.csv.gz
```

**Comment by bpbond** [comment]
Closing in prep for git bfg. Details recorded via #867 .

**Pull request closed without merging**